File: signing.go
package pact

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"
)

// SignedRequestHeaders are the custom HTTP headers Pact adds to requests.
const (
	// HeaderAgentIdentity carries the agent's identity ID.
	HeaderAgentIdentity = "X-Pact-Identity"

	// HeaderDelegationChain carries the base64url-encoded delegation chain JSON.
	HeaderDelegationChain = "X-Pact-Chain"

	// HeaderSignatureInput describes what components are covered by the signature.
	HeaderSignatureInput = "Signature-Input"

	// HeaderSignature carries the request signature.
	HeaderSignature = "Signature"

	// HeaderContentDigest carries the SHA-256 digest of the request body.
	HeaderContentDigest = "Content-Digest"
)

// coveredComponents are the HTTP message components included in the signature.
// Per RFC 9421, these are the fields that the signature binds to the request.
var coveredComponents = []string{
	"@method",
	"@path",
	"@authority",
	"date",
	HeaderAgentIdentity,
	HeaderDelegationChain,
}

// SignRequest signs an HTTP request with the agent's private key and attaches
// the delegation chain. This is what agents call before making any HTTP request.
//
// After signing:
//   - X-Pact-Identity header contains the agent's identity
//   - X-Pact-Chain header contains the delegation chain
//   - Signature-Input describes what's covered
//   - Signature contains the Ed25519 signature
//   - Content-Digest contains SHA-256 of the body (if present)
func SignRequest(req *http.Request, agent *Identity, chain DelegationChain) error {
	if agent.PrivateKey == nil {
		return fmt.Errorf("pact: cannot sign request — agent has no private key")
	}

	// Set the Date header if not already set
	if req.Header.Get("Date") == "" {
		req.Header.Set("Date", time.Now().UTC().Format(http.TimeFormat))
	}

	// Set agent identity header
	req.Header.Set(HeaderAgentIdentity, agent.ID)

	// Serialize and set delegation chain
	chainJSON, err := json.Marshal(chain)
	if err != nil {
		return fmt.Errorf("pact: failed to serialize delegation chain: %w", err)
	}
	req.Header.Set(HeaderDelegationChain, encodeBase64URL(chainJSON))

	// Compute content digest for request body (if present)
	if req.Body != nil && req.ContentLength > 0 {
		// Note: for requests with bodies, the caller should set Content-Digest
		// before signing, or we compute it here if possible.
		// For streaming bodies, Content-Digest must be pre-set.
	}

	// Build signature base (RFC 9421 §2.5)
	components := coveredComponents
	sigBase := buildSignatureBase(req, components)

	// Build Signature-Input header value
	sigInput := buildSignatureInput(components, agent.ID)
	req.Header.Set(HeaderSignatureInput, sigInput)

	// Include Signature-Input in what we sign (prevents tampering with input spec)
	sigBase += fmt.Sprintf("\"signature-input\": %s\n", sigInput)

	// Sign the signature base
	sig, err := agent.Sign([]byte(sigBase))
	if err != nil {
		return err
	}
	req.Header.Set(HeaderSignature, fmt.Sprintf("pact=:%s:", encodeBase64URL(sig)))

	return nil
}

// buildSignatureBase constructs the signature base string per RFC 9421 §2.5.
// This is the canonical string that gets signed.
func buildSignatureBase(req *http.Request, components []string) string {
	var sb strings.Builder

	for _, comp := range components {
		value := resolveComponent(req, comp)
		sb.WriteString(fmt.Sprintf("\"%s\": %s\n", strings.ToLower(comp), value))
	}

	return sb.String()
}

// resolveComponent extracts the value of a covered component from the request.
func resolveComponent(req *http.Request, component string) string {
	switch component {
	case "@method":
		return req.Method
	case "@path":
		path := req.URL.Path
		if path == "" {
			path = "/"
		}
		if req.URL.RawQuery != "" {
			path += "?" + req.URL.RawQuery
		}
		return path
	case "@authority":
		return req.Host
	default:
		// Regular header
		return req.Header.Get(component)
	}
}

// buildSignatureInput constructs the Signature-Input header value per RFC 9421.
func buildSignatureInput(components []string, keyID string) string {
	// Format: pact=("@method" "@path" "@authority" "date" ...);keyid="sha256:...";created=...
	var parts []string
	for _, comp := range components {
		parts = append(parts, fmt.Sprintf("\"%s\"", strings.ToLower(comp)))
	}

	created := time.Now().UTC().Unix()
	return fmt.Sprintf("pact=(%s);keyid=\"%s\";created=%d;alg=\"ed25519\"",
		strings.Join(parts, " "), keyID, created)
}

// ComputeContentDigest computes the SHA-256 digest of a request body.
// Returns the value for the Content-Digest header per RFC 9530.
func ComputeContentDigest(body []byte) string {
	hash := sha256.Sum256(body)
	return fmt.Sprintf("sha-256=:%s:", encodeBase64URL(hash[:]))
}

// RequestSignatureData holds the parsed signature data extracted from an HTTP request.
// Used by the verification pipeline.
type RequestSignatureData struct {
	// SignatureBase is the canonical string that was signed.
	SignatureBase string

	// Signature is the raw Ed25519 signature bytes.
	Signature []byte

	// AgentID is the claimed agent identity.
	AgentID string

	// Chain is the parsed delegation chain.
	Chain DelegationChain

	// SignatureInput is the parsed Signature-Input header.
	SignatureInput string

	// Timestamp is when the request was signed.
	Timestamp time.Time
}

// ExtractSignatureData parses the Pact signature data from an HTTP request.
// This is the first step of provider-side verification.
func ExtractSignatureData(req *http.Request) (*RequestSignatureData, error) {
	// Extract agent identity
	agentID := req.Header.Get(HeaderAgentIdentity)
	if agentID == "" {
		return nil, fmt.Errorf("pact: missing %s header", HeaderAgentIdentity)
	}

	// Extract and decode delegation chain
	chainEncoded := req.Header.Get(HeaderDelegationChain)
	if chainEncoded == "" {
		return nil, fmt.Errorf("pact: missing %s header", HeaderDelegationChain)
	}
	chainJSON, err := decodeBase64URL(chainEncoded)
	if err != nil {
		return nil, fmt.Errorf("pact: invalid delegation chain encoding: %w", err)
	}
	var chain DelegationChain
	if err := json.Unmarshal(chainJSON, &chain); err != nil {
		return nil, fmt.Errorf("pact: invalid delegation chain JSON: %w", err)
	}

	// Extract signature
	sigHeader := req.Header.Get(HeaderSignature)
	if sigHeader == "" {
		return nil, fmt.Errorf("pact: missing %s header", HeaderSignature)
	}
	sigBytes, err := parseSignatureHeader(sigHeader)
	if err != nil {
		return nil, err
	}

	// Extract Signature-Input
	sigInput := req.Header.Get(HeaderSignatureInput)
	if sigInput == "" {
		return nil, fmt.Errorf("pact: missing %s header", HeaderSignatureInput)
	}

	// Rebuild signature base for verification
	sigBase := buildSignatureBase(req, coveredComponents)
	sigBase += fmt.Sprintf("\"signature-input\": %s\n", sigInput)

	// Parse timestamp from Date header
	dateStr := req.Header.Get("Date")
	var timestamp time.Time
	if dateStr != "" {
		timestamp, _ = time.Parse(http.TimeFormat, dateStr)
	}

	return &RequestSignatureData{
		SignatureBase:  sigBase,
		Signature:      sigBytes,
		AgentID:        agentID,
		Chain:          chain,
		SignatureInput: sigInput,
		Timestamp:      timestamp,
	}, nil
}

// parseSignatureHeader extracts the raw signature bytes from the Signature header.
// Format: pact=:base64url_signature:
func parseSignatureHeader(header string) ([]byte, error) {
	// Expected format: pact=:base64url:
	prefix := "pact=:"
	if !strings.HasPrefix(header, prefix) {
		return nil, fmt.Errorf("pact: invalid Signature header format: %q", header)
	}
	rest := header[len(prefix):]
	if !strings.HasSuffix(rest, ":") {
		return nil, fmt.Errorf("pact: invalid Signature header format: %q", header)
	}
	encoded := rest[:len(rest)-1]
	return decodeBase64URL(encoded)
}

File: sdk/python/pact/signing.py
"""HTTP request signing for the Pact protocol.

Signs HTTP requests with the agent's private key and attaches the delegation
chain. Uses a subset of RFC 9421 HTTP Message Signatures.
"""

from __future__ import annotations

import hashlib
import json
import time as time_mod
from email.utils import formatdate
from typing import Any

from .delegation import Delegation, DelegationChain
from .encoding import decode_base64url, encode_base64url
from .identity import Identity


# HTTP headers used by Pact
HEADER_AGENT_IDENTITY = "X-Pact-Identity"
HEADER_DELEGATION_CHAIN = "X-Pact-Chain"
HEADER_SIGNATURE_INPUT = "Signature-Input"
HEADER_SIGNATURE = "Signature"
HEADER_CONTENT_DIGEST = "Content-Digest"

# Covered components in the signature base (RFC 9421)
COVERED_COMPONENTS = [
    "@method",
    "@path",
    "@authority",
    "date",
    HEADER_AGENT_IDENTITY,
    HEADER_DELEGATION_CHAIN,
]


class HttpRequest:
    """A simple HTTP request representation for signing/verification.

    This abstracts away the HTTP library (requests, httpx, aiohttp, etc.)
    so the signing code doesn't depend on any specific library.
    """

    def __init__(
        self,
        method: str,
        url: str,
        headers: dict[str, str] | None = None,
        body: bytes | None = None,
    ):
        self.method = method.upper()
        self.headers = {k: v for k, v in (headers or {}).items()}
        self.body = body

        # Parse URL
        from urllib.parse import urlparse
        parsed = urlparse(url)
        self.host = parsed.hostname or ""
        if parsed.port and parsed.port not in (80, 443):
            self.host = f"{parsed.hostname}:{parsed.port}"
        self.path = parsed.path or "/"
        if parsed.query:
            self.path += f"?{parsed.query}"

    def get_header(self, name: str) -> str:
        """Get a header value (case-insensitive)."""
        for k, v in self.headers.items():
            if k.lower() == name.lower():
                return v
        return ""

    def set_header(self, name: str, value: str) -> None:
        """Set a header value."""
        self.headers[name] = value


def sign_request(
    req: HttpRequest,
    agent: Identity,
    chain: DelegationChain,
) -> None:
    """Sign an HTTP request with the agent's private key and attach the delegation chain.

    After signing, the request will have:
    - X-Pact-Identity: the agent's identity ID
    - X-Pact-Chain: base64url-encoded delegation chain JSON
    - Signature-Input: describes what components are covered
    - Signature: the Ed25519 signature
    - Date: current UTC time (if not already set)

    Args:
        req: The HTTP request to sign.
        agent: The agent identity (must have private key).
        chain: The delegation chain.

    Raises:
        ValueError: If the agent has no private key.
    """
    if agent.private_key is None:
        raise ValueError("Cannot sign request — agent has no private key")

    # Set Date header if not already set
    if not req.get_header("Date"):
        req.set_header("Date", formatdate(
            timeval=None, localtime=False, usegmt=True))

    # Set agent identity header
    req.set_header(HEADER_AGENT_IDENTITY, agent.id)

    # Serialize and set delegation chain
    chain_dicts = [d.to_dict() for d in chain]
    chain_json = json.dumps(chain_dicts, separators=(",", ":")).encode("utf-8")
    req.set_header(HEADER_DELEGATION_CHAIN, encode_base64url(chain_json))

    # Build signature base
    sig_base = _build_signature_base(req, COVERED_COMPONENTS)

    # Build Signature-Input header value
    sig_input = _build_signature_input(COVERED_COMPONENTS, agent.id)
    req.set_header(HEADER_SIGNATURE_INPUT, sig_input)

    # Include Signature-Input in what we sign
    sig_base += f'"signature-input": {sig_input}\n'

    # Sign the signature base
    sig = agent.sign(sig_base.encode("utf-8"))
    req.set_header(HEADER_SIGNATURE, f"pact=:{encode_base64url(sig)}:")


def compute_content_digest(body: bytes) -> str:
    """Compute the SHA-256 digest of a request body.

    Returns the Content-Digest header value per RFC 9530.
    """
    h = hashlib.sha256(body).digest()
    return f"sha-256=:{encode_base64url(h)}:"


def _build_signature_base(req: HttpRequest, components: list[str]) -> str:
    """Build the signature base string per RFC 9421 §2.5."""
    parts = []
    for comp in components:
        value = _resolve_component(req, comp)
        parts.append(f'"{comp.lower()}": {value}\n')
    return "".join(parts)


def _resolve_component(req: HttpRequest, component: str) -> str:
    """Extract the value of a covered component from the request."""
    if component == "@method":
        return req.method
    elif component == "@path":
        return req.path if req.path else "/"
    elif component == "@authority":
        return req.host
    else:
        # Regular header
        return req.get_header(component)


def _build_signature_input(components: list[str], key_id: str) -> str:
    """Build the Signature-Input header value per RFC 9421."""
    parts = [f'"{comp.lower()}"' for comp in components]
    created = int(time_mod.time())
    return f'pact=({" ".join(parts)});keyid="{key_id}";created={created};alg="ed25519"'


def extract_signature_data(req: HttpRequest) -> dict[str, Any]:
    """Parse the Pact signature data from an HTTP request.

    This is the first step of provider-side verification.

    Returns:
        Dict with keys: signature_base, signature, agent_id, chain, timestamp.

    Raises:
        ValueError: If required headers are missing or malformed.
    """
    # Extract agent identity
    agent_id = req.get_header(HEADER_AGENT_IDENTITY)
    if not agent_id:
        raise ValueError(f"Missing {HEADER_AGENT_IDENTITY} header")

    # Extract and decode delegation chain
    chain_encoded = req.get_header(HEADER_DELEGATION_CHAIN)
    if not chain_encoded:
        raise ValueError(f"Missing {HEADER_DELEGATION_CHAIN} header")
    chain_json = decode_base64url(chain_encoded)
    chain_dicts = json.loads(chain_json)
    chain = [Delegation.from_dict(d) for d in chain_dicts]

    # Extract signature
    sig_header = req.get_header(HEADER_SIGNATURE)
    if not sig_header:
        raise ValueError(f"Missing {HEADER_SIGNATURE} header")
    sig_bytes = _parse_signature_header(sig_header)

    # Extract Signature-Input
    sig_input = req.get_header(HEADER_SIGNATURE_INPUT)
    if not sig_input:
        raise ValueError(f"Missing {HEADER_SIGNATURE_INPUT} header")

    # Rebuild signature base for verification
    sig_base = _build_signature_base(req, COVERED_COMPONENTS)
    sig_base += f'"signature-input": {sig_input}\n'

    # Parse timestamp from Date header
    date_str = req.get_header("Date")
    timestamp = None
    if date_str:
        from email.utils import parsedate_to_datetime
        try:
            timestamp = parsedate_to_datetime(date_str)
        except Exception:
            pass

    return {
        "signature_base": sig_base,
        "signature": sig_bytes,
        "agent_id": agent_id,
        "chain": chain,
        "signature_input": sig_input,
        "timestamp": timestamp,
    }


def _parse_signature_header(header: str) -> bytes:
    """Extract the raw signature bytes from the Signature header.

    Format: pact=:base64url_signature:
    """
    prefix = "pact=:"
    if not header.startswith(prefix):
        raise ValueError(f"Invalid Signature header format: {header!r}")
    rest = header[len(prefix):]
    if not rest.endswith(":"):
        raise ValueError(f"Invalid Signature header format: {header!r}")
    encoded = rest[:-1]
    return decode_base64url(encoded)
